File: Backend/src/main/java/com/webbandogiadung/backend/service/ProductService.java
package com.webbandogiadung.backend.service;

import com.webbandogiadung.backend.dto.ProductResponse;
import com.webbandogiadung.backend.exception.ResourceNotFoundException;
import com.webbandogiadung.backend.repository.ProductRepository;
import java.util.List;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional(readOnly = true)
public class ProductService {

    private final ProductRepository productRepository;
    private final ProductMapper productMapper;

    public ProductService(ProductRepository productRepository, ProductMapper productMapper) {
        this.productRepository = productRepository;
        this.productMapper = productMapper;
    }

    public List<ProductResponse> searchProducts(String category, String keyword) {
        return productRepository.search(blankToNull(category), blankToNull(keyword)).stream()
                .map(productMapper::toResponse)
                .toList();
    }

    public ProductResponse getProduct(Long id) {
        return productRepository.findDetailedById(id)
                .map(productMapper::toResponse)
                .orElseThrow(() -> new ResourceNotFoundException("Khong tim thay san pham id=" + id));
    }

    public List<String> getCategories() {
        return productRepository.findDistinctCategories();
    }

    private String blankToNull(String value) {
        return value == null || value.isBlank() ? null : value.trim();
    }
}

File: Backend/Services/ProductService.cs
using Backend.Dtos;
using Backend.Exceptions;
using Backend.Models;
using Backend.Repositories;

namespace Backend.Services;

public sealed class ProductService(IProductRepository productRepository)
{
    public IReadOnlyList<ProductResponse> SearchProducts(string? category, string? keyword) =>
        productRepository.Search(category, keyword).Select(ToResponse).ToList();

    public ProductResponse GetProduct(long id)
    {
        var product = productRepository.GetById(id)
            ?? throw new ResourceNotFoundException($"Khong tim thay san pham id={id}");

        return ToResponse(product);
    }

    public IReadOnlyList<string> GetCategories() => productRepository.GetCategories();

    public ProductResponse CreateProduct(ProductRequest request) =>
        ToResponse(productRepository.Add(ToProduct(request)));

    public ProductResponse UpdateProduct(long id, ProductRequest request)
    {
        var product = productRepository.Update(id, ToProduct(request))
            ?? throw new ResourceNotFoundException($"Khong tim thay san pham id={id}");

        return ToResponse(product);
    }

    public void DeleteProduct(long id)
    {
        if (!productRepository.Delete(id))
        {
            throw new ResourceNotFoundException($"Khong tim thay san pham id={id}");
        }
    }

    private static Product ToProduct(ProductRequest request) => new()
    {
        Category = request.Category.Trim(),
        Name = request.Name.Trim(),
        Price = request.Price,
        OldPrice = request.OldPrice,
        Rating = request.Rating,
        Sold = request.Sold,
        Stock = request.Stock,
        Image = request.Image.Trim(),
        Description = request.Description.Trim(),
        Gallery = request.Gallery?.Select(item => item.Trim()).Where(item => item.Length > 0).ToList() ?? [],
        Badges = request.Badges?.ToList() ?? [],
        Specs = request.Specs?.ToList() ?? []
    };

    private static ProductResponse ToResponse(Product product) => new(
        product.Id,
        product.Category,
        product.Name,
        product.Price,
        product.OldPrice,
        product.Rating,
        product.Sold,
        product.Stock,
        product.Image,
        product.Gallery,
        product.Badges,
        product.Description,
        product.Specs);
}
